### PR TITLE
Cleanup frozen FunctionGraph

### PR DESCRIPTION
**Pull request by ricardoV94**
Related to #2033 

* Removed special work-around for broken scalar signature (there are some fundamental issues described in #2044 but they are not specific to here)
* Intern and fix equality for FunctionGraph with constant outputs
* Add output node client to outputs (so it won't look like a variable has no clients, or if shared, only has the non-output clients).

### CODE REVIEW

File: pytensor/scalar/basic.py [hunk 1]
```diff
@@ -958,10 +958,31 @@ class ScalarVariable(_scalar_py_operators, Variable):
...
+class ScalarConstantSignature(tuple):
+    """Signature for ScalarConstant that handles NaN equality and hashing."""
+
+    def __eq__(self, other):
+        if type(self) is not type(other):
+            return False
+        (t0, d0), (t1, d1) = self, other
+        if t0 != t1:
+            return False
+        return (d0 == d1) or (np.isnan(d0) and np.isnan(d1))
```
Comment (jessegrabowski) [inline pytensor/scalar/basic.py:970]: are there any other corner cases with np.inf? I don't think so just checking.

File: tests/graph/test_fg.py [hunk 2]
```diff
@@ -903,8 +891,6 @@ def test_interned_constant_in_variables(self):
         cache hit overwrites memo[c2]=c1, c2 is evicted from variables
         while the cache-miss node still references it.
         """
-        from pytensor.graph.fg import FrozenFunctionGraph
```
Comment (jessegrabowski) [inline tests/graph/test_fg.py:906]: we should consider adding the ruff rule that all imports need to be on top of the file, im embarrassed i checked this in.

File: tests/graph/test_fg.py [hunk 3]
```diff
@@ -925,6 +911,33 @@ def test_interned_constant_in_variables(self):
             for inp in node.inputs:
                 assert inp in fg.variables
 
+    def test_constant_output_equality(self):
+        """FFGs with distinct but equal constant outputs should be equal."""
+        c1 = ScalarConstant(float64, 3.14)
```
Comment (jessegrabowski) [inline tests/graph/test_fg.py:916]: There are other circle constants you know: https://www.youtube.com/watch?v=dUGFb_HgG1c

File: pytensor/graph/fg.py [hunk 4]
```diff
@@ -992,38 +991,44 @@ def _resolve_input(inp, memo=memo):
...
+        # Create dummy Output nodes for each output, mirroring FunctionGraph.
+        # (It also makes eq/hash cheaper)
+        output_nodes = []
+        for i, o in enumerate(outputs):
+            try:
+                resolved = memo[o]
+            except KeyError:
+                if not isinstance(o, AtomicVariable):
+                    raise ValueError(
+                        f"Output variable {o} could not be mapped to a frozen graph variable. "
```
Comment (jessegrabowski) [inline pytensor/graph/fg.py:1003]: Is it really possible for a user to reach here? I ask because the error isn't really actionable, so it's another pytensor-special cryptic error if you do end up here somewhere.

**Review by jessegrabowski — APPROVED** [review]

File: pytensor/graph/fg.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/graph/fg.py:1003]: I had this doubt myself before. I don't think so, it would trigger in the first loop. Wanna take it out? 

One edge case is a NominalVariable output not specified as input. That should fail, which means this branch should only really accept Constants (not general AtomicVariables)

That may be findable only here but would need to check

Comment (jessegrabowski) [reply pytensor/graph/fg.py:1003]: We can make it an assert if we just want it for safety but dont think it should ever be hit (keeping the message)

**Comment by jessegrabowski** [comment]
Does this one need anything else?

File: pytensor/scalar/basic.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/scalar/basic.py:970]: np.inf is safe

**Comment by ricardoV94** [comment]
There was a bit of a nit of yours about the error message, but I went ahead since you didn't remember it MUHAHAHAH
